framework/src/transfers/MultiAppProjectionTransfer.C

```diff
@@ -220,44 +223,96 @@ MultiAppProjectionTransfer::assembleL2To(EquationSystems & es, const std::string
-  std::vector<MeshFunction *> from_fns(n_apps, NULL);
-  std::vector<MeshTools::BoundingBox *> from_bbs(n_apps, NULL);
-
-  // get bounding box, mesh function and solution for each subapp
-  for (unsigned int i = 0; i < n_apps; i++)
+  /********************
+  In order to assemble the system we need to evaluate the sub app solutions at
+  quadrature points in the master domain.  This is complicated by the fact that
+  each app is on its own MPI communicator.
+
+  Each processor will
+  1. Generate bounding boxes and mesh functions for it's local sub apps
+  2. Send the bounding boxes to the other processors in an all-to-all comm
+  3. Check its local quadrature points in the master domain to see which
+     bounding boxes they lie in
+  4. Send quadrature points to the processors with the containing bounding boxes
+  5. Recieve quadrature points from other processors, evaluate its mesh
+     functions at those points, and send the values back to the proper processor
+  6. Recieve mesh function evaluations from all relevant processors and decide
+     which one to use at every quadrature point (the lowest global app index
+     always wins)
+  7. And assemble the L2 system at its local quadrature points
+  ********************/
+
+  /********************
+  First step, get the bounding boxes and mesh functions for all the sub apps on
+  this processor.
+  ********************/
+  const unsigned int n_global_apps = _multi_app->numGlobalApps();
+  const unsigned int n_local_apps = _multi_app->numLocalApps();
+  std::vector<NumericVector<Number> *> serialized_from_solutions(n_local_apps, NULL);
+  std::vector<MeshFunction *> local_meshfuns(n_local_apps, NULL);
+  std::vector<MeshTools::BoundingBox> local_bboxes(n_local_apps);
+  std::vector<std::pair<Point, Point> > bb_points(n_local_apps);
+
+  unsigned int i_local = 0;
+  for (unsigned int i_global = 0; i_global < n_global_apps; i_global++)
   {
-    if (!_multi_app->hasLocalApp(i))
+    if (!_multi_app->hasLocalApp(i_global))
       continue;
 
-    MPI_Comm swapped = Moose::swapLibMeshComm(_multi_app->comm());
-
-    FEProblem & from_problem = *_multi_app->appProblem(i);
+    // Get the bounding box.
+    FEProblem & from_problem = *_multi_app->appProblem(i_global);
     EquationSystems & from_es = from_problem.es();
     MeshBase & from_mesh = from_es.get_mesh();
-    MeshTools::BoundingBox * app_box = new MeshTools::BoundingBox(MeshTools::processor_bounding_box(from_mesh, from_mesh.processor_id()));
-    from_bbs[i] = app_box;
+    MeshTools::BoundingBox app_box = MeshTools::BoundingBox(MeshTools::processor_bounding_box(from_mesh, from_mesh.processor_id()));
+
+    // Cast the bounding box into a pair of points to simplify MPI
+    // communication.  Translate the bounding box to the app's position.
+    bb_points[i_local] = static_cast<std::pair<Point, Point> >(app_box);
+    bb_points[i_local].first = bb_points[i_local].first + _multi_app->position(i_global);
```

Comment:
Good job thinking about needing to offset based on the App's position.

I've actually been thinking about adding a `MultiApp::appBoundingBox(unsigned int)` method to `MultiApp` that would give you back the properly translated bounding box.  It would simplify the need to do this kind of stuff in each `Transfer`.  What do you think?